Tools/autotest/common.py

```diff
@@ -1467,117 +1467,131 @@ def wait_seconds(self, seconds_to_wait):
-                      (alt_min, alt_max))
-        last_wait_alt_msg = 0
-        while self.get_sim_time_cached() < tstart + timeout:
-            m = self.mav.recv_match(type='GLOBAL_POSITION_INT', blocking=True)
-            if m is None:
-                continue
-            if relative:
-                alt = m.relative_alt/1000.0 # mm -> m
-            else:
-                alt = m.alt/1000.0 # mm -> m
-
-            climb_rate = alt - previous_alt
-            previous_alt = alt
-            if self.get_sim_time_cached() - last_wait_alt_msg > 1:
-                self.progress("Wait Altitude: Cur:%.02f min_alt:%.02f climb_rate: %.02f"
-                              % (alt, alt_min, climb_rate))
-                last_wait_alt_msg = self.get_sim_time_cached()
-            if alt >= alt_min and alt <= alt_max:
-                self.progress("Altitude OK")
+        def get_altitude(alt_relative=False, timeout2=30):
+            msg = self.mav.recv_match(type='GLOBAL_POSITION_INT', blocking=True, timeout=timeout2)
+            if msg:
+                if alt_relative:
+                    return msg.relative_alt / 1000.0  # mm -> m
+                else:
+                    return msg.alt / 1000.0  # mm -> m
+
+        def validator(value2, target2):
+            if altitude_min <= value2 <= altitude_max:
                 return True
-        raise WaitAltitudeTimout("Failed to attain altitude range")
+            else:
+                return False
+
+        self.wait_and_maintain("Altitude", altitude_min, lambda: get_altitude(relative, timeout), accuracy=(altitude_max - altitude_min), validator=lambda value2, target2: validator(value2, target2), maintain_target_time=maintain_target_time, timeout=timeout, called_function=called_function)
 
-    def wait_groundspeed(self, gs_min, gs_max, timeout=30):
+    def wait_groundspeed(self, speed_min, speed_max, maintain_target_time=None, timeout=30):
         """Wait for a given ground speed range."""
-        self.progress("Waiting for groundspeed between %.1f and %.1f" %
-                      (gs_min, gs_max))
-        last_print = 0
-        tstart = self.get_sim_time()
-        while self.get_sim_time_cached() < tstart + timeout:
-            m = self.mav.recv_match(type='VFR_HUD', blocking=True)
-            if self.get_sim_time_cached() - last_print > 1:
-                self.progress("Wait groundspeed %.3f, target:%.3f" %
-                              (m.groundspeed, gs_min))
-                last_print = self.get_sim_time_cached()
-            if m.groundspeed >= gs_min and m.groundspeed <= gs_max:
+        assert speed_min <= speed_max, "Minimum speed should be less than maximum speed."
+
+        def get_groundspeed(timeout2):
+            msg = self.mav.recv_match(type='VFR_HUD', blocking=True, timeout=timeout2)
+            if msg:
+                return msg.groundspeed
+
+        def validator(value2, target2):
+            if speed_min <= value2 <= speed_max:
```

Comment:
Simplify this.

Reply:
? that is the simplier form already... maybe you mean to expand it to two if ?

Reply:
`return speed_min <= value2 <= speed_max`